daml-lf/validation/src/test/scala/com/digitalasset/daml/lf/validation/TypingSpec.scala

```diff
@@ -302,156 +302,274 @@ class TypingSpec extends WordSpec with TableDrivenPropertyChecks with Matchers {
     }
 
     "reject ill formed terms" in {
-      val testCases = Table(
-        "non-well formed expression",
+
+      // In the following test cases we use the variable `nothing` when we
+      // cannot built an expression of the expected type. In those cases we
+      // expect the type checker to fail with the error we are testing before
+      // it tries to type check the variable `nothing`.
+      // Expressions of type τ, where τ has kind ⋆ → ⋆, are examples of
+      // such expressions that cannot be constructed.
+
+      val testCases = Table[Expr, PartialFunction[ValidationError, _]](
+        "non-well formed expression" -> "error",
         // ExpDefVar
-        E"x",
+        E"⸨ x ⸩" -> //
+          { case _: EUnknownExprVar => },
         // ExpApp
-        E"Λ (τ₁: ⋆) (τ₂ : ⋆). λ (e₁ : τ₂) (e₂ : τ₁) → (( e₁ e₂ ))",
-        E"Λ (τ₁: ⋆) (τ₂ : ⋆) (τ₃ : ⋆). λ (e₁ : τ₂ → τ₃) (e₂ : τ₁) → (( e₁ e₂ ))",
-        E"Λ (τ₁: ⋆) (τ₂ : ⋆) (τ₃ : ⋆). λ (e₁ : (τ₁ → τ₂) → τ₃) (e₂ : τ₁) → (( e₁ e₂ ))",
-        E"Λ (τ₁: ⋆) (τ₂ : ⋆) (τ₃ : ⋆). λ (e₁ : Bool) (e₂ : τ₃) → (( e₁ e₂ ))",
+        E"Λ (τ₁: ⋆) (τ₂ : ⋆). λ (e₁ : τ₂) (e₂ : τ₁) → ⸨ e₁ e₂ ⸩" -> //
+          { case _: EExpectedFunctionType => },
+        E"Λ (τ₁: ⋆) (τ₂ : ⋆) (τ₃ : ⋆). λ (e₁ : τ₂ → τ₃) (e₂ : τ₁) → ⸨ e₁ e₂ ⸩" -> //
+          { case _: ETypeMismatch => },
+        E"Λ (τ₁: ⋆) (τ₂ : ⋆) (τ₃ : ⋆). λ (e₁ : (τ₁ → τ₂) → τ₃) (e₂ : τ₁) → ⸨ e₁ e₂ ⸩" -> //
+          { case _: ETypeMismatch => },
+        E"Λ (τ₁: ⋆) (τ₂ : ⋆) (τ₃ : ⋆). λ (e₁ : Bool) (e₂ : τ₃) → ⸨ e₁ e₂ ⸩" -> //
+          { case _: EExpectedFunctionType => },
         // ExpTyApp
-        E"Λ (τ : ⋆ → ⋆) (σ: ⋆ → ⋆). λ (e : ∀ (α : ⋆). σ α) → (( e @τ ))",
-        E"Λ (τ : ⋆) (σ: ⋆ → ⋆). λ (e : ∀ (α : ⋆ → ⋆). σ α) → (( e @τ ))",
+        E"Λ (τ : ⋆ → ⋆) (σ: ⋆ → ⋆). λ (e : ∀ (α : ⋆). σ α) → ⸨ e @τ ⸩" -> //
+          { case _: EKindMismatch => },
+        E"Λ (τ : ⋆) (σ: ⋆). λ (e : σ) → ⸨ e @τ ⸩" -> //
+          { case _: EExpectedUniversalType => },
         // ExpAbs
-        E"Λ  (τ : ⋆ → ⋆) (σ: ⋆) . λ (e: τ → σ) → λ (x : τ) → (( e x ))",
+        E"⸨ λ (x : List) → () ⸩" -> //
+          { case _: EKindMismatch => },
         // ExpLet
```

Comment:
I think we should add a test here that the scoping rules do not allow for recursive let bindings:

``
 E"Λ  (τ : ⋆). ⸨ let x : τ = x in nothing ⸩" -> //
           { case _ : EUnknownExprVar => },```

Reply:
I prefer to add that in the test which is relative to variable scope, See lines [268-302](https://github.com/digital-asset/daml/pull/7905/files#R278-R302)

Reply:
👍 Sounds good, thanks!